driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -213,25 +227,41 @@ private Throwable checkOutFailed(final Throwable t) {
         Throwable result = t;
         if (t instanceof MongoTimeoutException) {
             connectionPoolListener.connectionCheckOutFailed(new ConnectionCheckOutFailedEvent(serverId, Reason.TIMEOUT));
-        } else if (t instanceof MongoOpenConnectionInternalException) {
+        } else if (t instanceof MongoOpenConnectionInternalException || t instanceof MongoConnectionPoolClearedException) {
```

Comment:
The invocation of the `unwrap` method cannot be moved out of `checkOutFailed` because its part of the logic of handling a checkout failure (the sole purpose of `MongoOpenConnectionInternalException` is to convey the failure reason). I split the `else if` such that `MongoOpenConnectionInternalException.unwrap` is now called only for `MongoOpenConnectionInternalException` arguments.

Reply:
If we're going to do that, then can we just remove the unwrap method and call `getCause()` directly?  (Note there is similar usage in `doMaintenance`)

Reply:
Changed in e750aaa3de537add3486abda229de911170f7c7c. It is ugly either way, but maybe this time it is less confusing.